### PR TITLE
Add authorizer to check whether a given user is allowed to perform a set of operations.

### PR DESCRIPTION
**Pull request by fabricematrat**

### CODE REVIEW

File: macaroonbakery/identity.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+
+class SimpleIdentity(Identity, ACLIdentity):
+    ''' A simple form of identity where the user is represented by a string.
+    '''
+    def __init__(self, user):
+        self._identity = user
+
+    def domain(self):
+        ''' Returning the empty domain.
```
Comment (rogpeppe) [inline macaroonbakery/identity.py]: "A simple identity has no domain"?

File: macaroonbakery/identity.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        '''
+        return ""
+
+    def id(self):
+        '''Return the simple identity as user
```
Comment (rogpeppe) [inline macaroonbakery/identity.py]: Return the user name as the id.

?

File: macaroonbakery/identity.py [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+        '''
+        return self._identity
+
+    def allow(self, acls):
+        '''Allowing the identity access to ACL members that are equal to id.
```
Comment (rogpeppe) [inline macaroonbakery/identity.py]: Allow access to any ACL members that was equal to the user name.
?

File: macaroonbakery/authorizer.py [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            return None, None
+        allowed = []
+        for op in ops:
+            acl = self.get_acl(op)
+            here = self.allow_public() and _is_public_acl(acl)
+            print(here)
+            if identity is not None:
+                try:
+                    allowed.append(identity.allow(acl))
+                except AttributeError:
```
[the remark this answers is not in the document]
Comment (rogpeppe) [reply macaroonbakery/authorizer.py]: Oh I see what's going on - you're checking whether the identity implements ACLIdentity.
Why not check directly before calling allow? That's why abstract base classes exist, right?

File: macaroonbakery/authorizer.py [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+            if fcaveats is not None:
+                caveats.extend(fcaveats)
+        return allowed, caveats
+
+
+class ACLAuthorizer(Authorizer):
+    ''' ACLAuthorizer is an Authorizer implementation that will check access
+    control list (ACL) membership of users. It uses get_acl to find out
+    the ACLs that apply to the requested operations and will authorize an
+    operation if an ACL contains the group "everyone" or if an ACLIdentity
```
Comment (rogpeppe) [inline macaroonbakery/authorizer.py]: or if the identity is an instance of ACLIdentity and its allow method returns True for the ACL. ?

File: macaroonbakery/authorizer.py [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+import abc
+
+# EVERYONE is recognized by ACLAuthorizer as the name of a
+# group that has everyone in it.
+EVERYONE = "everyone"
+
+
+class AuthorizerError(Exception):
+    """ AuthorizerError is the exception raised when when the authorization
+    cannot be.
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/authorizer.py]: Removed.

File: macaroonbakery/authorizer.py [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def authorize(self, id, ops):
+        ''' Checks whether the given identity (which will be None when there is
+        no authenticated user) is allowed to perform the given operations.
+        It should raise an exception only when the authorization cannot be
+        determined, not when the user has been denied access.
+
+        On success, each element of allowed holds whether the respective element
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/authorizer.py]: commented

File: macaroonbakery/authorizer.py [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+        raise NotImplementedError('users must define authorize to use this base'
+                                  ' class')
+
+
+class AuthorizerFunc(Authorizer):
+    ''' Implements a simplified version of Authorizer that operates on a single
+    operation at a time.
+    '''
+    def __init__(self, f):
+        self._f = f
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/authorizer.py]: done.

File: macaroonbakery/authorizer.py (hunk 5 above)
Comment (fabricematrat) [reply macaroonbakery/authorizer.py]: done.

File: macaroonbakery/authorizer.py [hunk 9]
```diff
@@ -0,0 +1,115 @@
...
+
+
+class ACLAuthorizer(Authorizer):
+    ''' ACLAuthorizer is an Authorizer implementation that will check access
+    control list (ACL) membership of users. It uses get_acl to find out
+    the ACLs that apply to the requested operations and will authorize an
+    operation if an ACL contains the group "everyone" or if an ACLIdentity
+    and its allow method returns True for the ACL.
+    '''
+    def __init__(self, allow_public, get_acl):
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/authorizer.py]: done.

File: macaroonbakery/authorizer.py [hunk 10]
```diff
@@ -0,0 +1,115 @@
...
+        self._allow_public = allow_public
+        self._get_acl = get_acl
+
+    def allow_public(self):
+        ''' If _allow_public is True and an ACL contains "everyone",
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/authorizer.py]: removed.

File: macaroonbakery/authorizer.py [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+        then authorization will be granted even if there is
+        no logged in user.
+        :return:
+        '''
+        return self._allow_public
+
+    def get_acl(self, op):
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/authorizer.py]: removed

File: macaroonbakery/authorizer.py [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+
+    def authorize(self, identity, ops):
+        '''Implements Authorizer.authorize by calling identity.allow to
+        determine whether the identity is a member of the ACLs associated with
+        the given operations.
+        '''
+        if ops is None or len(ops) == 0:
+            # Anyone is allowed to do nothing.
+            return None, None
+        allowed = []
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/authorizer.py]: done but have look @rogpeppe

File: macaroonbakery/authorizer.py [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+        for op in ops:
+            acl = self.get_acl(op)
+            here = self.allow_public() and _is_public_acl(acl)
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/authorizer.py]: done.

File: macaroonbakery/authorizer.py [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+            print(here)
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/authorizer.py]: removed.

File: macaroonbakery/authorizer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/authorizer.py]: use isinstance instead

File: macaroonbakery/authorizer.py [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+                try:
+                    allowed.append(identity.allow(acl))
+                except AttributeError:
+                    allowed.append(here)
+            else:
+                allowed.append(here)
+        return allowed, None
+
+
+def _is_public_acl(acls):
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/authorizer.py]: done.

File: macaroonbakery/identity.py [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+class Identity(object):
+    ''' Holds identity information declared in a first party caveat added when
+    discharging a third party caveat.
+    '''
+    __metaclass__ = abc.ABCMeta
+
+    @abc.abstractmethod
+    def id(self):
+        ''' Returns the id of the user
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/identity.py]: done.

File: macaroonbakery/identity.py [hunk 17]
```diff
@@ -0,0 +1,71 @@
...
+
+        May be an opaque blob with no human meaning. An id is only considered
+        to be unique with a given domain.
+        '''
+        raise NotImplementedError('users must define id to use this base'
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/identity.py]: done.

File: macaroonbakery/identity.py [hunk 18]
```diff
@@ -0,0 +1,71 @@
...
+                                  ' class')
+
+    @abc.abstractmethod
+    def domain(self):
+        '''Return the domain of the user.
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/identity.py]: done.

File: macaroonbakery/identity.py [hunk 19]
```diff
@@ -0,0 +1,71 @@
...
+
+        This will be empty if the user was authenticated
```
[the remark this answers is not in the document]
Comment (fabricematrat) [reply macaroonbakery/identity.py]: empty !

File: macaroonbakery/identity.py (hunk 1 above)
Comment (fabricematrat) [reply macaroonbakery/identity.py]: done.

File: macaroonbakery/identity.py (hunk 2 above)
Comment (fabricematrat) [reply macaroonbakery/identity.py]: done.

File: macaroonbakery/identity.py (hunk 3 above)
Comment (fabricematrat) [reply macaroonbakery/identity.py]: done.

File: macaroonbakery/authorizer.py [hunk 20]
```diff
@@ -0,0 +1,99 @@
...
+                                  'subclass')
+
+
+class AuthorizerFunc(Authorizer):
+    ''' Implements a simplified version of Authorizer that operates on a single
+    operation at a time.
+    '''
+    def __init__(self, f):
+        '''
+        :param f: a function that takes an identity that operates on a single
```
Comment (rogpeppe) [inline macaroonbakery/authorizer.py]: This needs to say what the function returns too - I'd have no idea how to use this if I read the docs.
Also, it's not clear from this description that it has two arguments.

File: macaroonbakery/authorizer.py [hunk 21]
```diff
@@ -0,0 +1,99 @@
...
+class ACLAuthorizer(Authorizer):
+    ''' ACLAuthorizer is an Authorizer implementation that will check access
+    control list (ACL) membership of users. It uses get_acl to find out
+    the ACLs that apply to the requested operations and will authorize an
+    operation if an ACL contains the group "everyone" or if the identity is
+    an instance of ACLIdentity and its allow method returns True for the ACL.
+    '''
+    def __init__(self, get_acl, allow_public=False):
+        '''
+        :param get_acl return the ACL that applies to the given operation.
```
Comment (rogpeppe) [inline macaroonbakery/authorizer.py]: get_acl will be called with an Op and should return the ACL that applies (an array of string ids).

?

File: macaroonbakery/authorizer.py [hunk 22]
```diff
@@ -0,0 +1,99 @@
...
+    def authorize(self, identity, ops):
+        '''Implements Authorizer.authorize by calling identity.allow to
+        determine whether the identity is a member of the ACLs associated with
+        the given operations.
+        '''
+        if ops is None or len(ops) == 0:
+            # Anyone is allowed to do nothing.
+            return None, None
+        allowed = [False] * len(ops)
+        for i, op in enumerate(ops):
```
Comment (rogpeppe) [inline macaroonbakery/authorizer.py]: How about this?

    hasAllow = isinstance(identity, ACLIdentity)
    for i, op in enumerate(ops):
        acl = self._get_acl(op)
        if hasAllow:
            allowed[i] = identity.allow(acl)
        else:
             allowed[i] = (self._allow_public || ident is not None) and EVERYONE in acl

**Pull request merged**